src/propagate_precision.cpp

```diff
@@ -0,0 +1,191 @@
+#include <migraphx/propagate_precision.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/eliminate_convert.hpp>
+#include <unordered_set>
+#include <unordered_map>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace {
+struct precision
+{
+    shape::type_t type;
+
+    friend bool operator==(const precision& xp, const precision& yp) { return xp.type == yp.type; }
+    friend bool operator<(const precision& xp, const precision& yp)
+    {
+        bool is_less = false;
+        shape::visit(xp.type, [&](auto x) {
+            shape::visit(yp.type, [&](auto y) {
+                if(x.is_integral() != y.is_integral())
+                    return;
```

Comment:
For readability i think it is better to write it as
`if(not x.is_integral() or not y.is_integral())`